File: Tim/Data/UserData.java
package Tim.Data;

import Tim.Tim;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.sql.Types;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.UUID;

public class UserData {
	// Data for throttling lookups
	public boolean registrationDataRetrieved = false;
	public Date lastWhoisCheck;

	// Non-synced reference data
	public HashSet<String> nicks = new HashSet<>();
	public HashMap<UUID, Integer> recordedWars = new HashMap<>();

	// DB synced data
	public UUID id;
	public String authedUser;
	public int totalSprintWordcount;
	public int totalSprints;
	public int totalSprintDuration;
	public boolean raptorAdopted;
	public String raptorName;
	public String raptorFavoriteColor;
	public int raptorBunniesStolen;
	public Date lastBunnyRaid;

	public void save() {
		try (Connection con = Tim.db.pool.getConnection(3000)) {
			PreparedStatement s = con.prepareStatement("REPLACE INTO `users` (`id`, `authed_user`, `total_sprint_wordcount`, `total_sprints`, `total_sprint_duration`, `raptor_adopted`, `raptor_name`, `raptor_favorite_color`, `raptor_bunnies_stolen`, `last_bunny_raid`) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?);");

			s.setString(1, id.toString());
			s.setString(2, authedUser);
			s.setInt(3, totalSprintWordcount);
			s.setInt(4, totalSprints);
			s.setInt(5, totalSprintDuration);
			s.setBoolean(6, raptorAdopted);
			s.setString(7, raptorName);
			s.setString(8, raptorFavoriteColor);
			s.setInt(9, raptorBunniesStolen);
			if (lastBunnyRaid != null) {
				s.setDate(10, new java.sql.Date(lastBunnyRaid.getTime()));
			} else {
				s.setNull(10, Types.DATE);
			}

			s.executeUpdate();
		} catch (SQLException ex) {
			Tim.printStackTrace(ex);
		}
	}
}

File: timmy/data/userdata.py
import uuid

from typing import Optional, Set

from timmy import db_access


class UserData:
    registration_data_retrieved: bool
    last_whois_check: Optional[float]
    nicks: Set[str]
    recorded_wars: Set[str]
    authed_user: str
    global_admin: bool
    total_sprint_wordcount: int
    total_sprints: int
    total_sprint_duration: float
    raptor_adopted: bool
    raptor_name: str
    raptor_favorite_color: str
    raptor_bunnies_stolen: int
    last_bunny_raid: Optional[float]

    def __init__(self):
        self.registration_data_retrieved = False
        self.last_whois_check = None

        self.nicks = set()
        self.recorded_wars = set()

        self.uuid = None
        self.authed_user = ""
        self.global_admin = False
        self.total_sprint_wordcount = 0
        self.total_sprints = 0
        self.total_sprint_duration = 0
        self.raptor_adopted = False
        self.raptor_name = ""
        self.raptor_favorite_color = ""
        self.raptor_bunnies_stolen = 0
        self.last_bunny_raid = None

    def save(self):
        update_statement = "REPLACE INTO `users` (`id`, `authed_user`, `global_admin`, `total_sprint_wordcount`, " \
                           "`total_sprints`, `total_sprint_duration`, `raptor_adopted`, `raptor_name`, " \
                           "`raptor_favorite_color`, `raptor_bunnies_stolen`, `last_bunny_raid`) VALUES " \
                           "(%(uuid)s, %(authed_user)s, %(global_admin)s, %(total_sprint_wordcount)s, " \
                           "%(total_sprints)s, %(total_sprint_duration)s, %(raptor_adopted)s, %(raptor_name)s, " \
                           "%(raptor_favorite_color)s, %(raptor_bunnies_stolen)s, %(last_bunny_raid)s)"

        connection = db_access.connection_pool.get_connection()
        cursor = connection.cursor()
        cursor.execute(update_statement, {
            'uuid':                   str(self.uuid),
            'authed_user':            self.authed_user,
            'global_admin':           self.global_admin,
            'total_sprint_wordcount': self.total_sprint_wordcount,
            'total_sprints':          self.total_sprints,
            'total_sprint_duration':  self.total_sprint_duration,
            'raptor_adopted':         self.raptor_adopted,
            'raptor_name':            self.raptor_name,
            'raptor_favorite_color':  self.raptor_favorite_color,
            'raptor_bunnies_stolen':  self.raptor_bunnies_stolen,
            'last_bunny_raid':        self.last_bunny_raid
        })

        connection.close()
